get s3 url via aws-sdk-go, fix #9189

williamjoy opened the pull request:
ref #9189 

Debug log with uploading to s3 `ap-northeast-1`:

```go
DBUG[09-13|03:27:02] Image rendered                           logger=png-renderer path=/usr/share/grafana/data/png/4epc3CbtT0MCKVeatDjw.png
DBUG[09-13|03:27:02] Uploading image to s3%!(EXTRA string=url = , string=https://s3-ap-northeast-1.amazonaws.com/grafana.codo/J4W7M82V7wnsUJ4HobD4.png) 
INFO[09-13|03:27:03] uploaded                                 logger=alerting.notifier url=https://s3-ap-northeast-1.amazonaws.com/grafana.codo/J4W7M82V7wnsUJ4HobD4.png
INFO[09-13|03:27:03] Sending notification                     logger=alerting.notifier type=slack id=1 isDefault=false
```

Looks good for slack
<img width="691" alt="screen shot 2017-09-13 at 15 36 58" src="https://user-images.githubusercontent.com/3284451/30365168-7a3f095e-9899-11e7-8c9b-4d8fba86ef41.png">

williamjoy committed:
get s3 url via aws-sdk-go, fix #9189

williamjoy commented:
link issue #9189

williamjoy commented:
Tested s3 upload to `cn-north-1` region works:
```
DBUG[09-13|04:36:02] Image rendered                           logger=png-renderer path=/usr/share/grafana/data/png/wDoM5ClGL1r1F4Er7Gfi.png
DBUG[09-13|04:36:02] Uploading image to s3%!(EXTRA string=url = , string=https://s3.cn-north-1.amazonaws.com.cn/grafana.codo/XNYjfNaBSjLa65LEjm4N.png) 
INFO[09-13|04:36:02] uploaded                                 logger=alerting.notifier url=https://s3.cn-north-1.amazonaws.com.cn/grafana.codo/XNYjfNaBSjLa65LEjm4N.png
```

TODO: update unit test

williamjoy commented:
Testing upload to `us-east-1`, the s3 url should be correct. ( Didn't setup s3 bucket in `us-east-1`)
```golang
Loading a web page: http://localhost:3000/dashboard-solo/db/test-alert?&panelId=1 status: success 30000
DBUG[09-13|22:10:02] Image rendered                           logger=png-renderer path=/usr/share/grafana/data/png/gui6kSAE4x3w7vFdtveU.png
DBUG[09-13|22:10:02] Uploading image to s3%!(EXTRA string=url = , string=https://s3.amazonaws.com/grafana.codo/f0DgN8JWwzT2RdBwkQS1.png) 
EROR[09-13|22:10:04] Failed to upload alert panel image.      logger=alerting.notifier error="BucketRegionError: incorrect region, the bucket is not in 'us-east-1' region\n\tstatus code: 301, request id: , host id: "
```

williamjoy commented:
@torkelo I've tested s3 upload to `us-east-1` `ap-northeast-1` `cn-north-1`, looks good for getting s3 url. Also tested slack notification works. Can you take a look?

bergquist commented:
Looks good.

My main concern is that this breaks backwards comp ability.

torkelo commented:
Yea, need to support the old option as well, otherwise upgrade will break this for all existing installs who have this configured

williamjoy committed:
config bucket and region for s3 uploader

this is to support cn-north-1 region as it can get s3 url programatically.
also keeps support 'bucket_url' for backward compatiblity

williamjoy commented:
yep, `bucket_url` config support added back. but `bucket` and `region` have higher priority than `bucket_url` config.

bergquist commented:
@williamjoy great! we will review this once we have time. Please be patient :)

williamjoy commented on pkg/components/imguploader/s3uploader.go:
```diff
@@ -77,10 +80,5 @@ func (u *S3Uploader) Upload(imageDiskPath string) (string, error) {
-		return "https://" + u.bucket + ".s3.amazonaws.com/" + key, nil
-	} else {
-		return "https://" + u.bucket + ".s3-" + u.region + ".amazonaws.com/" + key, nil
-	}
```
if `u.bucket` contains `.`, e.g. `foo.bar`, the return url **was** `https://foo.bar.s3-us-west-2.amazonaws.com/`, it also causing slack does not render the image attachment , because the SSL certs common name is `*.s3-us-west-2.amazonaws.com` , which does **not** match `foo.bar.s3-us-west-2.amazonaws.com`

williamjoy commented:
@bergquist sure :)

torkelo commented on conf/defaults.ini:
```diff
@@ -458,7 +458,8 @@ url = https://grafana.com
 provider =
 
 [external_image_storage.s3]
-bucket_url =
```
this needs to be added back, as the keys defined in defaults are used to know what keys can be set (and what ENV variables to look for, so it it is not here you cannot override it from command line or ENV variable)

torkelo replied:
since it was such a minor fix I merged and did this myself, thanks for contributing!

williamjoy commented:
Thanks , good to know that!
